lib/ansible/module_utils/csharp/Ansible.Become.cs

```diff
@@ -1,690 +1,870 @@
+                using (process)
                 {
-                    IntPtr hToken = IntPtr.Zero;
-                    // According to CreateProcessWithTokenW we require a token with
-                    //  TOKEN_QUERY, TOKEN_DUPLICATE and TOKEN_ASSIGN_PRIMARY
-                    // Also add in TOKEN_IMPERSONATE so we can get an impersontated token
-                    TokenAccessLevels desired_access = TokenAccessLevels.Query |
-                        TokenAccessLevels.Duplicate |
-                        TokenAccessLevels.AssignPrimary |
-                        TokenAccessLevels.Impersonate;
-
-                    if (OpenProcessToken(hProcess, desired_access, out hToken))
+                    using (SafeNativeHandle hProcess = NativeMethods.OpenProcess(accessFlags, false, (UInt32)process.Id))
                     {
-                        string sid = GetTokenUserSID(hToken);
-                        if (sid == "S-1-5-18")
+                        if (hProcess.IsInvalid)
+                            continue;
+
+                        SafeNativeHandle hToken;
+                        NativeMethods.OpenProcessToken(hProcess, dwAccess, out hToken);
+                        if (hToken.IsInvalid)
+                            continue;
+
+                        using (hToken)
                         {
-                            CloseHandle(hProcess);
-                            return hToken;
+                            if (!sid.Equals(GetTokenUserSID(hToken)))
+                                continue;
+
+                            // Filter out any Network logon tokens, using become with that is useless when S4U
+                            // can give us a Batch logon
+                            NativeHelpers.SECURITY_LOGON_TYPE tokenLogonType = GetTokenLogonType(hToken);
+                            if (tokenLogonType == NativeHelpers.SECURITY_LOGON_TYPE.Network)
+                                continue;
+
+                            // Check that the required privileges are on the token
+                            if (requiredPrivileges != null)
+                            {
+                                List<string> actualPrivileges = GetTokenPrivileges(hToken);
+                                int missing = requiredPrivileges.Where(x => !actualPrivileges.Contains(x)).Count();
+                                if (missing > 0)
+                                    continue;
+                            }
+
+                            SafeNativeHandle dupToken;
+                            if (!NativeMethods.DuplicateTokenEx(hToken, TokenAccessLevels.MaximumAllowed,
+                                IntPtr.Zero, NativeHelpers.SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation,
+                                NativeHelpers.TOKEN_TYPE.TokenPrimary, out dupToken))
+                            {
+                                continue;
+                            }
+
+                            return dupToken;
```

Comment:
I'm not sure it actually matters, but you went to some trouble to dispose the `Process` objects as you're looping over them. This short-circuit prevents all the processes you didn't hit yet in the loop from being disposed. At a quick glance, it doesn't look like Process has a finalizer, so maybe it does matter (at least if this code were to run lots of times in a long-lived process)?

Reply:
Found that Component (which Process inherits) has a finalizer https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Primitives/src/System/ComponentModel/Component.cs#L23 so the handles are still being disposed off when GC gets to it.